### PR TITLE
Implement "on_no_change" for files

### LINKED ISSUE #670: add a not_changed hook to resources, similar to on_change but only gets executed if the resource does not change
For example something like this:

``` perl
file "/path/to/file",
  source => "files/local/file",
  not_changed => sub {
    say "nothing was changed...";
  };
```

### PR DESCRIPTION
**Pull request by choroba**
This PR fixes #670 by implementing the proposed feature. I changed the name to "on_no_change" as it sounds much better than "on_unchanged", but English is my second language, so my feeling might be wrong.

## Checklist

- [x] based on top of latest source code 
- [x] changelog entry included 
- [x] tests pass on Travis CI 
- [ ] git history is clean    
- [ ] git commit messages are [well-written](https://chris.beams.io/posts/git-commit/#seven-rules)

### CODE REVIEW

**Comment by choroba** [comment]
Greetings from the PullRequest.Club.

**Comment by ferki** [comment]
Thanks for your contribution and participation in Pull Request Club, @choroba! ;)

Yes, I think `on_no_change` fits better the existing interface than `not_changed` or `on_unchanged`, as it is closer to its `on_change` pair.

I expected the first implementation would only consider the `file` command. Thanks for going the extra mile and adding this option to other related commands as well, even though it inflated the scope. I'll give it a full review at a later time.

**Comment by ferki** [comment]
I'm rebasing this PR on top of current default branch due to the current merge conflicts. I'm also curious how the code coverage changes with this. We might need to add further tests accordingly.

File: lib/Rex/Commands/File.pm [hunk 1]
```diff
@@ -1061,7 +1071,7 @@ sub _append_or_update {
   my ( $new_line, @m );
 
   # check if parameters are in key => value format
-  my ( $option, $on_change );
+  my ( $option, $on_change, $on_no_change );
```
Comment (ferki) [inline lib/Rex/Commands/File.pm:1074]: Let's mention the on_no_change option in the docs of this function too.

**Review by ferki — CHANGES_REQUESTED** [review]
Thanks again @choroba for your contribution!

As I commented earlier, I already rebased it on top of current default branch. I had to add an autogenerated perltidy-only commit but I didn't have to change anything else.

Based on my full reading of the changes, there's only slight changes required:
- add some more docs about the on_no_change hook
- fix one debug message

I plan to take care of that soon, so it could be already included in the next release. I might or might not squash together some of the commits once this is finished, but I'll try to make sure not to change the original commits if possible.

File: lib/Rex/Commands/File.pm [hunk 2]
```diff
@@ -740,6 +741,11 @@ sub file {
   if ( $__ret->{changed} == 1 && $on_change_done == 0 ) {
     &$on_change($file);
   }
+  elsif ( $__ret->{changed} == 0 ) {
+    Rex::Logger::debug(
+      "File $file has not been changed... Running on_no_change");
+    ( $option->{on_no_change} || sub { } )->($file);
```
[the remark this answers is not in the document]
Comment (ferki) [reply lib/Rex/Commands/File.pm]: I moved it in a follow up.

File: lib/Rex/Commands/File.pm [hunk 3]
```diff
@@ -1130,32 +1145,46 @@ sub _append_or_update {
...
+    }
+    elsif ($on_no_change) {
       $new_md5 ||= "";
 
-      Rex::Logger::debug("File $file has been changed... Running on_change");
-      Rex::Logger::debug("old: $old_md5");
-      Rex::Logger::debug("new: $new_md5");
-      &$on_change($file);
+      Rex::Logger::debug(
+        "File $file has been changed (md5 $new_md5)... Running on_no_change");
```
[the remark this answers is not in the document]
Comment (ferki) [reply lib/Rex/Commands/File.pm]: Fixed.

File: lib/Rex/Commands/File.pm (hunk 1 above)
Comment (ferki) [reply lib/Rex/Commands/File.pm:1079]: I've included a few more mentions.

**Review by ferki — APPROVED** [review]

**Comment by ferki** [comment]
Oh, last minute notice: this does not completely fix #670 yet, as there are a few more resources which have `on_change` hooks, but it's a great start, and probably one of the most important one anyway. So after merging, I probably will need to reopen the issue, and make further notes there.

**Comment by choroba** [comment]
Thanks for solving all the review comments before I had time to even look at them :-)

**Pull request merged**